### DISCUSSION TITLE
POST ciphers => error 422 : missing field `type`

### DISCUSSION THREAD

**Discussion by tomitomas** [dani-garcia/vaultwarden#4793] [Q&A] [answered]
Hi there, 

first of all : thanks for the amazing job you've done.
I've been using VaultWarden for 1 years, and it's a amazing tools ! ❤️ 

Today, I'm trying to use another (usefull ?) tools in complement of VW => [goldwarden](https://github.com/quexten/goldwarden/). 
The main aim of that tools is to communicate with a Bitwarden instance (VaultWarden?) to save and exchange data to allow connexion to server using ssh keys. 

When I try to add a new key on that tool, I get an 422 error. 
Having a look at the VW log we can see a `missing field 'type'` error. 

```
[2024-07-28 09:24:46.329][request][INFO] POST /api/ciphers
[2024-07-28 09:24:46.331][vaultwarden::api::core::ciphers::_][WARN] Data guard `Json < CipherData >` failed: 

{
    "Type": 2,
    "Name": "2.r1YDWm6ff5E+4+PeNH+Ayg==|/teiq9po6K6G+uYH4WwCdg==|JBJjQs8guL0Frj0/iwA9nrLFACWcNvc+uvbINfHqdXQ=",
    "Edit": false,
    "RevisionDate": "0001-01-01T00:00:00Z",
    "DeletedDate": "0001-01-01T00:00:00Z",
    "Fields": [{
            "Type": 0,
            "Name": "2.oVGTpCyF8S0YePLvmwbU+w==|Csw63hPEmQ2Wk89fC3up/Q==|g4PKZFvb8HnbxD7G3JDfQamAWXpKi06bgU2APtmN/wQ=",
            "Value": "2.ecHZZdFb5wbC8bSZQGFfQA==|h1mxmNjX9A2F1Ux7noThxQ==|BM+EqaehU9YvsDm+p9cJF/Xv+7aTUvwlL/eTKmIUbp8="
        }, {
            "Type": 0,
            "Name": "2.eQpKR1ShL0umeYLCjtVwmw==|Nfa34X40Hzw1NACKqeXD4g==|cH8DqtM5F8JwnEsD8a/I196dkfdgXRFwlcwjh3jBsRc=",
            "Value": "2.2Cl5Dacfa33PDgNpRcDAWA==|5mzwA8KjI4b4HMOpJZr347BEkHkf+4YzH2JDxnlvA9qbR+7kAPs4at0bmqTzhekuyOEwYspKoThXKVIQEmveKbnC8EAk8QplBDL6HlFGa+5GMD0KTGbI+3ydQRs7sXkb|o2KNUdk0TKoFYAsZ/tef5C/U6kd/0mF/jha2998PY18="
        }, {
            "Type": 1,
            "Name": "2.HmmrnmgJBhaltLS3Ave2lw==|ehWxQM+Cty+mUOpAMkIZ5w==|7wwurQRNeFCZdcivekGiiaoB4O7p8v2/kgDevvH5rgU=",
            "Value": "2.lotSRNI0/phyzQq/r8kTzQ==|YBw0zVfwNixtLOOQlV2mZEgQbdWQxBrt47zWsETNI3jEW/FykB6il1av2yC8SUYKN9v1odRYJ0OSOsedXma+dmL9hHJU53D6mPlsl1QqhKUv9eM/0BGBg8ac4LiiyJFVK53uw2TsncHkNjncsbhaLT3SNeAIe5X+3lqJfhxU2s/26LLeDdDU/mUwmzg1h55W7HlivR3AhRkXZmIHK7dFgN6rt/rWBL7tGDEdrDVqvhEA/y7RwH2STLICz1ez7JL5lWe1MQoOtEeWtqmhHK4r6Iv3E/jQYOQJsz0+v1KrsOCQkjaZTUvv3I27FclDmHXE83pXbe2HB4Isqgl+srSqh2tmxSFgjeIXmE1UG6p8W631ZwmXLoYsTHMudT4+nkzAs4sKimOmIKMYnkApxf01/5o2ZWSg9eMJXnpUMkPCr0MW8QHrvdkqFb9+ISv2bZ9CgTpzMnr65MXk9QougSTgS1XtOxy3BClQ9OZZBQ+jg3Mc1WziEy3CcAJkXslfOuuDnx9h8MqUOrBsHADTKWmexw==|vltP+WTV1E78vgAVkSBOMH2abXYG3PiGu0rmO56ahIo="
        }
    ],
    "Notes": "2.2Cl5Dacfa33PDgNpRcDAWA==|5mzwA8KjI4b4HMOpJZr347BEkHkf+4YzH2JDxnlvA9qbR+7kAPs4at0bmqTzhekuyOEwYspKoThXKVIQEmveKbnC8EAk8QplBDL6HlFGa+5GMD0KTGbI+3ydQRs7sXkb|o2KNUdk0TKoFYAsZ/tef5C/U6kd/0mF/jha2998PY18=",
    "SecureNote": {
        "Type": 0
    }
}

Error("missing field `type`", line: 1, column: 1738)).
[2024-07-28 09:24:46.331][rocket::server::_][WARN] No 422 catcher registered. Using Rocket default.
```

it seems that the payload keys are coming in uppercase => which make the call fails on vaultwarden side. 

Is there anything I can do to make it workng ? 
Any help or idea will be appreciate

Thanks a lot !

**Comment by BlackDex** [comment]
Send the key in lowercase. Which actually should be done for Bitwarden too. In the past all the clients did a case-convert to allow a mixture of Pascal case and camelCase.

The new native clients of Bitwarden only allows camelCase.
And the clients also only send in that format.

**Reply by tomitomas** [reply]
thanks for the return

> Send the key in lowercase.

I won't be able to change the code myself :)



> The new native clients of Bitwarden only allows camelCase.

anywhere I can find this information ? 
to share it with the developper

**Reply by BlackDex** [reply]
Let them check our code and that of Bitwarden.

**Reply by tomitomas** [reply]
@quexten if by any chance you could have a look here, regarding issue https://github.com/quexten/goldwarden/issues/262 :)

**Reply by quexten** [reply, marked as answer]
Thanks for the ping. I agree, it sounds like this is not a bug in vaultwarden, but rather caused by fixes to support the mobile clients, making vaultwarden a bit stricter than the official server in what request format is accepted. Tracking the fix here https://github.com/quexten/goldwarden/pull/265 (yet untested)
